### DISCUSSION TITLE
Overlapping method route error. Nested routers

### DISCUSSION THREAD

**Discussion by Kimishu** [tokio-rs/axum#3062] [Q&A] [answered]
### Summary

Greetings! 
I don't know where I should ask this, in utoipa or here, so here it is. I've checked issues page for this and find something old, but..
I have a main router where I've described OpenAPI. Next I want to add a **whole group of routes**. It's like _"I have a routes that works with user data and I wanna merge it in one group - Users"_. And as a result, I created a different function that returns a router with routes. And I got error: `Overlapping method route. Handler for 'POST /' already exists`
```rust
//user_handler.rs
fn get_user_routes() -> OpenApiRoutes {
   OpenApiRouter::new().routes(routes!(
   test_post_1,
   test_post_2
   ))
}
...
//main.rs
let (router, api) = OpenApiRouter::with_openapi(ApiDoc::openapi())
.nest("/user", get_user_routes()) //and here I wanna add it.
```

In addition, there is a problem: If I change `.routes(routes!(...))` to `.route(...).route(...)` error disappears **with documentation**.
Thanks in advance! I hope I'll find solution.. 

### axum version

0.7.9

**Comment by mladedav** [comment]
I don't know utopia, but this sounds like what they're warning about in the panics section [here](https://docs.rs/utoipa-axum/latest/utoipa_axum/macro.routes.html#panics).

**Reply by Kimishu** [reply]
Thank you for reply! I've checked this and yeah, it's really what it is.
So, then, if axum works like that too, could you give me any idea how to group routes properly? And if its impossible now, should I create, for example, a request for it?

As I mention before, at the beginning, I want something like that. (Before I used rocket, where I can do that but in another things its much worse than axum, so, thanks for great framework!)
```
/user/create
/user/get/{id}
/user/update
```

**Reply by mladedav** [reply, marked as answer]
Sorry, I've never used that library so I can't really help you. From what I gathered skimming the docs, the `routes!` macro creates a `MethodRouter`, i.e. it should group just different handlers for a single route and then I assume you add each of these to the router.

But again, I really have no idea, it's just a guess, I think you'll have better luck asking there.

**Reply by Kimishu** [reply]
Okay, I will. Thank you again for help and your time!

**Reply by Kimishu** [reply]
By the way, what can u recommend for documentation? Maybe for axum exists something else, instead of utoipa?

**Reply by mladedav** [reply]
I personally didn't need open API in rust yet so I can't really tell. I think I've heard about aide before for that so you might want to check that out.

**Reply by juhaku** [reply]
> From what I gathered skimming the docs, the `routes!` macro creates a `MethodRouter`, i.e. it should group just different handlers for a single route and then I assume you add each of these to the router.

Yup, that is what it does. With `routes!` macro you can group handlers per HTTP method and path. Though you can call `routes!` as many times as you want. So it is not necessary to add all routes in single `routes!` call. The only limitation is that you cannot define multiple handlers per http method within single call. E.g. two or more `get` calls cannot be added in same `routes!` call and similarly more than one `post` call or other http method is not possible.
